refactor+fix(op-node/sequencing): l1 origin selection improvements

geoknee opened the pull request:
* refactors l1 origin selection code  (+89/-61) so that:
  -  core business logic is in a pure, well unit-tested function (without side effects or network calls),
  - save a few network calls by using the cache
  - optimize for readability, using sentinel errors to describe the behaviour more clearly
* modifies recover mode behavior around origin selection, to not try to match auto derivation if close to the l1 tip
* adds comprehensive unit test for the new pure fn (+220 lines)
* adds an action test for sequencing window expiry recovery (+130 lines)
* adds an acceptance test for recover mode, ensuring it disables itself appropriately when recovery is done (+60 lines)

**Note to reviewer**: although this PR adds a significant amount of code, this is mostly test code. The production code stayed around the same number of lines but I tried to make it much more readable / understandable (as well as a small modification to the actual behavior, which could be pinched off into it's own PR, which would be worth about 60 lines including the testing).

geoknee committed:
WIP

geoknee committed:
wip

geoknee committed:
WIP

geoknee committed:
Treat NotFound next L1 origin as chain end

geoknee committed:
Use recover mode in sequence window expiry test

geoknee committed:
Invoke fault proof earlier and fix typos

Run env.RunFaultProofProgram after computing l2SafeHead
(l2SafeHead.Number.Uint64()/2) and replace the FromGenesis call with
RunFaultProofProgram. Fix minor comment typos and wrap a long log line.

geoknee committed:
reduce diff

geoknee committed:
Use requireL1OriginAt helper in test

Replace manual error assertions around FindL1Origin with
requireL1OriginAt and remove the now-unused derive import.

geoknee committed:
Introduce L2Sequencer.ActMaybeL2StartBlock

geoknee committed:
add TestRecoverModeWhenChainHealthy acceptance test

sysgo only

geoknee committed:
Add SetSequencerRecoverMode and enable debug logs

geoknee committed:
Adjust L1 block time and sequence window test

Set default L1 block time to 12s to match action helper assumptions.
Increase sequencer window size in the test to 50. Compute drift from
UnsafeL2 headers (use UnsafeL2.L1Origin). Simplify L1 mining to always
BatchMineAndSync and remove the extra numL1Blocks > 10 lag guard.

geoknee committed:
restore stub

geoknee committed:
WIP

geoknee committed:
name errs

geoknee committed:
refactor

geoknee committed:
fix

geoknee committed:
add test

geoknee committed:
Rename error constant and add L1 origin tests

Rename ErrInvalidL1OriginChild to ErrNextL1OriginOrphaned and adjust the
error text. Add test cases covering L1 reorg and invalid L1 origin, and
refactor the test case construction in the unit test.

geoknee committed:
Use drift check for next L1 origin and update tests

Compute driftNext and use it to decide adoption of the next L1 origin
instead of comparing absolute timestamps. Bump MaxSequencerDrift to 1800
and add tests covering max-drift and negative-drift scenarios.

geoknee committed:
Refactor L1 origin selection and error handling

Delegate origin decision logic to FindL1OriginOfNextL2Block and handle
synchronous fetch when in recover mode. Remove recover-mode fetching
from CurrentAndNextOrigin and return cached values instead. Update
sequencer error handling to distinguish invalid/orphaned origin (which
triggers a reset) from temporary lookup errors.

geoknee committed:
fixes

geoknee committed:
fixes

geoknee committed:
lint

geoknee committed:
don't use pointers

saves some translation code

geoknee committed:
handle retries without a "temporary error"

fixes action tests

geoknee committed:
use Fjord drift constant

geoknee committed:
fix origin_selector_test

mainly just asserting on sentinel errors, and taking account of small
optimization (fewer network calls)

geoknee commented on op-node/rollup/sequencing/origin_selector.go:
```diff
@@ -70,80 +69,56 @@ func (los *L1OriginSelector) OnEvent(ctx context.Context, ev event.Event) bool {
+	// If the cache doesn't have the next origin, but we now
+	// know we definitely need it, fetch it and try again.
+	if errors.Is(err, ErrNextL1OriginRequired) {
+		nextOrigin, _ = los.fetch(ctx, currentOrigin.Number+1)
```
Chat with @sebastianst , why don't we inspect the error here?

geoknee commented on op-node/rollup/sequencing/origin_selector_2.go:
```diff
@@ -0,0 +1,67 @@
+		if matchAutoDerivation {
+			// This can cause unsafe block production to slow to the rate of L1 block production, if the L1 origin is caught up to the L1 Head.
+			// Code higher up the call stack should ensure that matchAutoDerivation is false under such conditions.
+			return eth.L1BlockRef{}, ErrNextL1OriginRequired
```
Will we ever hit this? If the surrounding code guarantees the next origin is provided?

geoknee replied:
Changed the parent function to not always provide the next origin.

geoknee committed:
Simplify FindL1Origin

geoknee committed:
move new pure function into a method on los

geoknee commented on op-node/rollup/sequencing/origin_selector.go:
```diff
@@ -238,3 +206,63 @@ func (los *L1OriginSelector) reset() {
+// FindL1OriginOfNextL2Block finds the L1 origin of the next L2 block.
+// It returns an error if there is no way to build a block satisfying
+// derivation constraints with the supplied data.
+// You can pass a nil pointer for l1OriginChild if it is not yet available
```
```suggestion
// You can pass an empty l1OriginChild if it is not yet available
```

geoknee replied:
[204f4a1](https://github.com/ethereum-optimism/optimism/pull/18589/commits/204f4a15d003fc2d3f20b72a9e987d020ced4ba2)

geoknee commented on op-node/rollup/sequencing/sequencer.go:
```diff
@@ -503,21 +503,23 @@ func (d *Sequencer) startBuildingBlock() {
-
-	if !(l2Head.L1Origin.Hash == l1Origin.ParentHash || l2Head.L1Origin.Hash == l1Origin.Hash) {
+	switch err {
+	case nil:
```
Good call out, I will fix this.

geoknee replied:
[057e394](https://github.com/ethereum-optimism/optimism/pull/18589/commits/057e394bf970b46e9668ab5103099879b3d0b36e)

geoknee committed:
Update comment to refer to empty nextL1Origin

geoknee committed:
Use errors.Is for L1 origin error checks

geoknee committed:
Return L1 origin on validation errors

geoknee committed:
Add expectedResult to origin selector tests

Set expectedResult for the test cases with l2Head d1000 and e1000 to
assert the expected L1 origins (c101 and c100 respectively)

axelKingsley commented on op-e2e/actions/helpers/utils.go:
```diff
@@ -15,7 +15,7 @@ func DefaultRollupTestParams() *e2eutils.TestParams {
 		SequencerWindowSize: 120,
 		ChannelTimeout:      120,
-		L1BlockTime:         15,
+		L1BlockTime:         12, // Many of the action helpers assume a 12s L1 block time
```
out-of-scope: this sounds like a value we should enshrine as an Action Test Const

geoknee replied:
Yes something to bear in mind when we implement action tests in Rust.

axelKingsley commented on op-e2e/actions/proofs/sequence_window_expiry_test.go:
```diff
@@ -5,26 +5,47 @@ import (
 
-// Run a test that proves a deposit-only block generated due to sequence window expiry.
+// Run a test that proves a deposit-only block generated due to sequence window expiry,
+// and then recovers the chain using sequencer recover mode.
```
Do you think we could make this a 2nd test? If we're looking at test output, and `SequenceWindowExpireTest` is failing, I would like to know if it's the deposit-only that's failing, or the recovery mode.

geoknee replied:
We could do that, but then the second test needs to repeat a lot of what the first test is doing (since we apply recover mode to a chain where the sequencing window has already expired). 

I think if the test fails, it will still be easy to know the reason because of the `require` statements. I will add some extra messages into these statements to help in the understanding of any test failure.

axelKingsley commented on op-node/rollup/sequencing/sequencer.go:
```diff
@@ -503,21 +503,23 @@ func (d *Sequencer) startBuildingBlock() {
-	}
-
-	if !(l2Head.L1Origin.Hash == l1Origin.ParentHash || l2Head.L1Origin.Hash == l1Origin.Hash) {
+	switch {
+	case err == nil:
+	case errors.Is(err, ErrInvalidL1Origin), errors.Is(err, ErrNextL1OriginOrphaned):
```
I think these switch/cases are nice looking, but I'm trying to decide if there is a functional difference in the behavior here, or if it's just stylistic update. Could you revert this file alone and have the rest of the PR function as-is?

The logic being removed from these early statements appear to be showing back up lower down as default cases.

geoknee replied:
There is a functional difference, because `d.l1OriginSelector.FindL1Origin()` is now responsible for detecting the "inconsistent l1 origin" case (e.g. from a reorg) whereas previously this function was doing that. There are new types of errors being returned which we need to respond to differently.

Personally I find using `if/else` when introspecting the specific error type can be a bit messy and require more indentation than a switch. For example, 1) is the error nil? 2) if the error is not nil, is it type a ? 3) if it is not type a, is it type b? 

To answer your question, it would not be possible to revert this file alone and have things function as is, because a reorg error would be treated as a temporary error, instead of causing a reset like we need it to.

But also, this PR is trying to aggressively optimise this code for readability and maintainability. So while it could be done in a similar way to the old code, I like this better. It should be easier to see that all types of error are being explicitly handled.

axelKingsley commented on op-node/rollup/sequencing/origin_selector.go:
```diff
@@ -238,3 +206,63 @@ func (los *L1OriginSelector) reset() {
+// This method is designed to be pure (it only reads the cfg property of the receiver)
+// and should not have any side effects.
+func (los *L1OriginSelector) findL1OriginOfNextL2Block(
+	l2Head eth.L2BlockRef,
+	currentL1Origin eth.L1BlockRef, nextL1Origin eth.L1BlockRef,
+	matchAutoDerivation bool) (eth.L1BlockRef, error) {
+
+	if (currentL1Origin == eth.L1BlockRef{}) {
+		panic("origin-selector: currentL1Origin is empty")
```
Could this ever happen when launching from genesis, launching from a dev-L1, or something like that?

geoknee replied:
Good question, I will l dig into this. Anyway, it seems prudent here to just return `ErrInvalidOrigin` instead of panicking.

geoknee replied:
Apparently we do on occasion also `reset` both `los.currentL1Origin` and `los.nextL1Origin`, so that's another code path that could potentially trip this condition.

geoknee replied:
So when launching from genesis, we set the l1 origin of the first sequenced L2 block (number `1`) to a hardcoded value in the rollup config:

https://github.com/ethereum-optimism/optimism/blob/ce9fa62d0c0325304fc37d91d87aa2e16a7f8356/op-node/rollup/derive/payload_util.go#L14-L25

regardless of whether the L1 is a dev network or not. This means that `los.CurrentAndNextOrigin` should always return a non-empty `currentL1Origin` to feed to `los.findL1OriginOfNextL2Block`.

I will add some comments to this effect, and keep the panic here.

axelKingsley commented on op-e2e/actions/proofs/sequence_window_expiry_test.go:
```diff
@@ -5,26 +5,47 @@ import (
+		p.MaxSequencerDrift = 1800 // use 1800 seconds (30 minutes), which is the protocol constant since Fjord
+	})
+
+	// It seems more difficult (almost impossible) to recover from sequencing window expiry with span batches,
+	// since the singular batches within are invalidated _atomically_.
+	// That is to say, if the oldest batch in the span batch fails the sequencing window check
+	// (l1 origin + seq window < l1 inclusion)
+	// All following batches are invalidated / dropped as well.
+	// https://github.com/ethereum-optimism/optimism/blob/73339162d78a1ebf2daadab01736382eed6f4527/op-node/rollup/derive/batches.go#L96-L100
```
I think this just means you'd do span-batches of size 1.

In fact, IIRC we chatted about this a while back, it is probably a good strategy during recovery to blast out batches of Size `1`. You can target the current block, and if you fail to get included, you adjust your transaction. It is wasteful on batcher funds, but allows for the chain operator to more directly resume control over the chain.

I wonder if calldata might be a good way to put the batch out, to avoid blob contention?

geoknee replied:
I think the current approach doesn't require the batcher to detect whether its transaction got included. By sending multiple batches in a channel, the inclusion time becomes less critical because it just means a little more or less of that channel ends up being discarded by the verifier nodes. And the bit that is not rejected is much larger (contains many more blocks) so the safe head will make a greater leap.

Calldata is probably a good shout, we probably don't want to wait to fill a blob with empty blocks (although there will be up to 10,000 L2 blocks we want to batch).

axelKingsley reviewed (approved):
Any concerns I'd have about the potential impact of this change are outweighed by the tests included (and the broad implicit testing the L1 Origin gets through all tests generally).

Very cleanly done, and I love the exhaustive tests on the functional unit.

I have some comments, but mostly they are out of scope.

geoknee committed:
Add assertion message and clarify origin comments

Provide a helpful failure message in sequence_window_expiry_test when
the safe head doesn't advance after the sequencing window expires.
Document that the current L1 origin should always be non-empty and add a
panic guard. Clarify the rationale for requiring the next L1 origin,
include a source link, and note the effect on unsafe block production.

geoknee committed:
Store recoverMode and add comment period

geoknee committed:
Update op-node/rollup/sequencing/origin_selector.go

Co-authored-by: almanax-ai[bot] <174396398+almanax-ai[bot]@users.noreply.github.com>

geoknee commented on op-node/rollup/sequencing/origin_selector.go:
```diff
@@ -70,80 +70,50 @@ func (los *L1OriginSelector) OnEvent(ctx context.Context, ev event.Event) bool {
-			return eth.L1BlockRef{}, fmt.Errorf("cannot build next L2 block past current L1 origin %s by more than sequencer time drift, and failed to find next L1 origin: %w", currentOrigin, err)
+	// Try to find the L1 origin given the current data in cache
+	o, err := los.findL1OriginOfNextL2Block(
+		l2Head,
```
[4e76749](https://github.com/ethereum-optimism/optimism/pull/18589/commits/4e76749c5f5c79135095b162772c35e1ded4b3f3)

geoknee committed:
Update op-node/rollup/sequencing/origin_selector.go

Co-authored-by: almanax-ai[bot] <174396398+almanax-ai[bot]@users.noreply.github.com>
